File: Python/32.longest-valid-parentheses.py
#
# @lc app=leetcode id=32 lang=python
#
# [32] Longest Valid Parentheses
#
# https://leetcode.com/problems/longest-valid-parentheses/description/
#
# algorithms
# Hard (26.61%)
# Likes:    2438
# Dislikes: 110
# Total Accepted:    225.7K
# Total Submissions: 848.1K
# Testcase Example:  '"(()"'
#
# Given a string containing just the characters '(' and ')', find the length of
# the longest valid (well-formed) parentheses substring.
# 
# Example 1:
# 
# 
# Input: "(()"
# Output: 2
# Explanation: The longest valid parentheses substring is "()"
# 
# 
# Example 2:
# 
# 
# Input: ")()())"
# Output: 4
# Explanation: The longest valid parentheses substring is "()()"
# 
# 
#

# @lc code=start
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        ans = 0
        dp = [0 for _ in range(len(s))]
        for i in range(1, len(s)):
            if s[i] == ')':
                if s[i-1] == '(':
                    dp[i] = dp[i-2] + 2

                elif i - dp[i-1] > 0 and s[i-dp[i-1]-1]=='(':
                    dp[i] = dp[i-1] + dp[i-dp[i-1]-2] + 2
            ans = max(ans, dp[i])
        return ans

        # 栈顶保存当前扫描的时候，合法序列前的一个位置位置下标是多少
        # ans = 0
        # stack = []
        # stack.append(-1)
        # for i in range(len(s)):
        #     if s[i] == '(':
        #         stack.append(i)
        #     else:
        #         stack.pop()
        #         if not stack:
        #             stack.append(i)
        #         else:
        #             ans = max(ans, i-stack[-1])     
        # return ans
# @lc code=end

sol = Solution()
s = "())((())))"
sol.longestValidParentheses(s)

File: C++/32.longest-valid-parentheses.cpp
/*
 * @lc app=leetcode.cn id=32 lang=cpp
 *
 * [32] Longest Valid Parentheses
 */
#include<string>
#include<vector>
#include<iostream>
using namespace std;

// @lc code=start
class Solution {
public:
    int longestValidParentheses(string s) {
        int ans = 0;
        vector<int> dp(s.size()+1, 0);

        for (int i=1; i<s.size(); i++) {
            if (s[i] == ')') {
                if (s[i-1] == '(')
                    dp[i+1] = dp[i-1]+2;
                else if (i-dp[i]>0 && s[i-dp[i]-1]=='(')
                    dp[i+1] = dp[i] + dp[i-dp[i]-1] +2;
            }
            ans = max(ans, dp[i+1]);
        }
        return ans;
    }
};
// @lc code=end

int main() {
    Solution sol;
    string s = "()(())";
    int ans = sol.longestValidParentheses(s);
    cout<<"ans="<<ans<<endl;
}
